sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzurePath.java

```diff
@@ -58,182 +119,422 @@ public Path getRoot() {
+        throws IOException {
+        throw new UnsupportedOperationException("WatchEvents are not supported.");
     }
 
     /**
+     * Unsupported.
+     * <p>
      * {@inheritDoc}
      */
     @Override
     public WatchKey register(WatchService watchService, WatchEvent.Kind<?>... kinds) throws IOException {
-        throw new UnsupportedOperationException();
+        throw new UnsupportedOperationException("WatchEvents are not supported.");
     }
 
     /**
+     * Unsupported
+     * <p>
      * {@inheritDoc}
      */
     @Override
     public Iterator<Path> iterator() {
-        return null;
+        return Flux.fromArray(this.splitToElements(this.withoutRoot()))
+            .map(s -> this.parentFileSystem.getPath(s))
+            .toIterable()
+            .iterator();
     }
 
     /**
+     * This result of this method is identical to a string comparison on the underlying path strings.
+     *
      * {@inheritDoc}
      */
     @Override
     public int compareTo(Path path) {
-        return 0;
+        if (!(path instanceof AzurePath)) {
+            throw Utility.logError(logger, new ClassCastException("Other path is not an instance of AzurePath."));
+        }
+
+        return this.pathString.compareTo(((AzurePath) path).pathString);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return this.pathString;
+    }
+
+    /**
+     * A path is considered equal to another path if it is associated with the same file system instance and if the
+     * path strings are equivalent.
+     *
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        AzurePath paths = (AzurePath) o;
+        return Objects.equals(parentFileSystem, paths.parentFileSystem)
+            && Objects.equals(pathString, paths.pathString);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(parentFileSystem, pathString);
+    }
+
+    // Used to ensure we only try to access containers that are mounted.
+    boolean validRoot(String fileStoreName) {
+        Boolean validRootName = Flux.fromIterable(parentFileSystem.getFileStores())
+            .map(FileStore::name)
+            .hasElement(fileStoreName)
+            .block();
+        return validRootName != null && validRootName;
+    }
+
+    private String withoutRoot() {
+        Path root = this.getRoot();
+        String str = this.pathString;
+        if (root != null) {
+            str = this.pathString.substring(root.toString().length());
+        }
+        if (str.startsWith(this.parentFileSystem.getSeparator())) {
+            str = str.substring(1);
+        }
+
+        return str;
+    }
+
+    private String[] splitToElements() {
+        return this.splitToElements(this.pathString);
```

Comment:
Given there is numerous calls to this method should the split pathString be stored as a field on the instance?

Reply:
I was wondering that as I wrote it. The consequence is that these objects basically double their memory usage, right? Since we would then store the pathString as one object and then all of its components again separately. I didn't feel like I had enough knowledge of customer scenarios to know if they were cpu or memory bound, so I was going to defer that decision until we hear one way or the other from customers unless you feel strongly about it. Does that make sense?

Reply:
That makes sense to me